### ISSUE TITLE
Bug: [no-useless-assignment] unexpected error when use a variable as the JSX element without any variable props

### ISSUE THREAD

**Issue by otomad** [typescript-eslint/typescript-eslint#10219]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&fileType=.tsx&code=KYDwDg9gTgLgBAE2AMwIYFcA29noHYDGMAlhHnAIJhgAUAlHAN4CwAUHHAWQM7xd4xgAuAF44Ac2AwAwmUED6AbjYd%2BvOLIGpieYFADqxGAAsAaqijFUAI0zAAClAhhuouACIExAG7vl7Th54TRhtXQMjYwBlUJICAHkoADkIR2dXMU8fPxU4KCl0KHIaXI4AHgA%2BUo44MpCwvUMTc0sbOzSXKoCamsZ%2BeRgAX2rygHp6nUbIlqtbBydOthHaifCm6NjiBOTUhe4unp7GdxjUOLhBEBh3Ye6asvG5BoiTU7jElI79pbvy1anXpttp89pxMKhuNwkqgALbAETuXhnLYXUDXOCjA73TG5Oj%2BW5sUCQWBwXCEEhkCRSEJCGD0Ji5Ow4CAQNzuGZtYCoq45AJqeD5bhYeBiZAs-wcMWszIAVW4wDskLgEO4xHEeDhAl5HHyMEK5EFwvxbCAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0pFpp9pJwYAL4ghQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export default function App() {
	const content = getContent();
	const ContainerWithVariableProps = "div";
	const ContainerWithStaticOrNoProps = "div";
	return (
		<>
			<ContainerWithVariableProps>{content}</ContainerWithVariableProps>

			<ContainerWithStaticOrNoProps>{"Static text"}</ContainerWithStaticOrNoProps>

			<ContainerWithStaticOrNoProps className="static text" />
		</>
	);
}

function getContent() {
	let foo = "Variable text";
	const result = foo;
	foo = "Useless assignment";
	return result;
}
```

### ESLint Config

```javascript
export default [
	tseslint.configs.base, // Comment this line will work fine.
	{
		files: ["**/*.{js,jsx,ts,tsx}"],
		rules: {
			"no-useless-assignment": "error",
		},
	},
];
```

### tsconfig

```jsonc
{
	"compilerOptions": {
		"allowJs": true,
		"jsx": "react-jsx",
	}
}
```

### Expected Result

`foo` in 23rd line should report the error "This assigned value is not used in subsequent statements",
but `ContainerWithStaticOrNoProps` in 4th line **should not** report the error "This assigned value is not used in subsequent statements".

### Actual Result

Both `foo` in 23rd line and `ContainerWithStaticOrNoProps` in 4th line report the error "This assigned value is not used in subsequent statements".

### Additional Info

1. This is a rule in ESLint, that is used in React JSX (eslint-plugin-react), but it is a TypeScript ESLint bug.
2. When disable TypeScript ESLint, it works fine.
3. Even if you use the `.jsx` file extension instead of the `.tsx` file extension, as long as TypeScript ESLint is enabled, the error will occur.
4. In https://github.com/eslint/eslint/issues/18984, it is mentioned that enabling `react/jsx-uses-vars` will make it work properly, but in fact it will only work without TypeScript ESLint, and it will still be abnormal after enabling TypeScript ESLint.

The above reproduced the issue using the TypeScript ESLint online playground,
and here is a minimal reproduction where you can see that enabling eslint-plugin-react still doesn't works properly:

https://stackblitz.com/edit/vitejs-vite-dmr77k?file=src%2FApp.jsx

Although StackBlitz does not support ESLint real-time error display, you can run the command `npm run lint` in the terminal below to get the message.

5. When use a variable as the JSX element, the variable will unexpected report the error "This assigned value is not used in subsequent statements". But if the JSX element has any **variable** props (or children), the error will not be reported. If all of the the props (including children) of the JSX element are **static** value, the error will still be reported. For example:

These will be unexpected reported:
```tsx
const Container = "div";
<Container />
<Container>Hello</Container>
<Container>{"Hello"}</Container>
<Container>{123}</Container>
<Container>{true}</Container>
<Container>{["foo"]}</Container>
<Container>{{ foo: "bar" }}</Container>
<Container className="hello" />
<Container className={"hello"} />
<Container hidden={true} />
<Container hidden />
```
These will not be reported, and will work properly:
```tsx
const foo = "bar";
const Container = "div";
<Container>{foo}</Container>
<Container>{[foo]}</Container>
<Container>{{ foo }}</Container>
<Container>{{ bar: foo }}</Container>
<Container className={foo} />
<Container className={foo}>Hello</Container>
<Container className="hello">{foo}</Container>
<Container className={foo}>{{ foo: "bar" }}</Container>
```

##### Debug Info
```
  eslint ./src/App.jsx --debug
  eslint:cli CLI args: [ './src/App.jsx', '--debug' ] +0ms
  eslint:cli Using flat config? true +2ms
  eslint:cli Running on files +4ms
  eslint:eslint Using config loader LegacyConfigLoader +0ms
  eslint:eslint Using file patterns: ./src/App.jsx +0ms
  eslint:eslint Deleting cache file at /home/projects/vitejs-vite-dmr77k/.eslintcache +1ms
  eslint:eslint 1 files found in: 1ms +1ms
  eslint:config-loader Calculating config for file /home/projects/vitejs-vite-dmr77k/src/App.jsx +0ms
  eslint:config-loader [Legacy]: Searching for eslint.config.js +0ms
  eslint:config-loader [Legacy]: Calculating config for /home/projects/vitejs-vite-dmr77k/src/App.jsx +2ms
  eslint:config-loader [Legacy]: Using config file /home/projects/vitejs-vite-dmr77k/eslint.config.js and base path /home/projects/vitejs-vite-dmr77k +0ms
  eslint:config-loader Calculating config array from config file /home/projects/vitejs-vite-dmr77k/eslint.config.js and base path /home/projects/vitejs-vite-dmr77k +0ms
  eslint:config-loader Loading config file /home/projects/vitejs-vite-dmr77k/eslint.config.js +1ms
  eslint:config-loader Loading config from /home/projects/vitejs-vite-dmr77k/eslint.config.js +0ms
  eslint:config-loader Config file URL is file:///home/projects/vitejs-vite-dmr77k/eslint.config.js +0ms
  eslint:rules Loading rule 'consistent-return' (remaining=289) +0ms
  eslint:rules Loading rule 'dot-notation' (remaining=288) +15ms
  eslint:rules Loading rule 'init-declarations' (remaining=287) +5ms
  eslint:rules Loading rule 'max-params' (remaining=286) +1ms
  eslint:rules Loading rule 'no-dupe-class-members' (remaining=285) +23ms
  eslint:rules Loading rule 'no-empty-function' (remaining=284) +4ms
  eslint:rules Loading rule 'no-invalid-this' (remaining=283) +10ms
  eslint:rules Loading rule 'no-loop-func' (remaining=282) +1ms
  eslint:rules Loading rule 'no-loss-of-precision' (remaining=281) +1ms
  eslint:rules Loading rule 'no-magic-numbers' (remaining=280) +1ms
  eslint:rules Loading rule 'no-restricted-imports' (remaining=279) +12ms
  eslint:rules Loading rule 'no-unused-expressions' (remaining=278) +62ms
  eslint:rules Loading rule 'no-useless-constructor' (remaining=277) +3ms
  eslint:rules Loading rule 'prefer-destructuring' (remaining=276) +10ms
  eslint:rules Loading rule 'no-useless-assignment' (remaining=275) +57ms
  eslint:eslint Lint /home/projects/vitejs-vite-dmr77k/src/App.jsx +1s
  eslint:linter Linting code for /home/projects/vitejs-vite-dmr77k/src/App.jsx (pass 1) +0ms
  eslint:linter Verify +0ms
  eslint:linter With flat config: /home/projects/vitejs-vite-dmr77k/src/App.jsx +1ms
  eslint:languages:js Parsing: /home/projects/vitejs-vite-dmr77k/src/App.jsx +0ms
  eslint:languages:js Parsing successful: /home/projects/vitejs-vite-dmr77k/src/App.jsx +26ms
  eslint:languages:js Scope analysis: /home/projects/vitejs-vite-dmr77k/src/App.jsx +0ms
  eslint:languages:js Scope analysis successful: /home/projects/vitejs-vite-dmr77k/src/App.jsx +0ms
  eslint:linter Generating fixed text for /home/projects/vitejs-vite-dmr77k/src/App.jsx (pass 1) +62ms
  eslint:source-code-fixer Applying fixes +0ms
  eslint:source-code-fixer shouldFix parameter was false, not attempting fixes +1ms

/home/projects/vitejs-vite-dmr77k/src/App.jsx
   4:8  error  This assigned value is not used in subsequent statements  no-useless-assignment
  23:2  error  This assigned value is not used in subsequent statements  no-useless-assignment

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by JoshuaKGoldberg** [comment]
This is an interesting bug. My hunch is it's probably something in https://typescript-eslint.io/packages/scope-manager and/or https://typescript-eslint.io/packages/typescript-estree. I haven't looked too closely. Accepting PRs!

**Comment by yeonjuan** [comment]
This is likely due to the fact that the typescript-eslint scope manager handles jsxIdentifier referencing, while the eslint scope manager does not.

For example, in the code below, eslint-scope manager recognizes only one reference for `AA` (Identifier), while typescript eslint scope-manager recognizes two references (Identifier, jsxIdentifier).

```jsx
function App() {
  const AA = () => <></>;
  return (
    <>
      <AA>
      </AA>
    </>
  );
}
```

- eslint scope manager
<img width="280" alt="스크린샷 2024-11-22 오전 1 22 18" src="https://github.com/user-attachments/assets/1772fe1b-b1bd-4b85-8c49-3d5b0c76875d">

- typescript-eslint scope manager
<img width="536" alt="스크린샷 2024-11-22 오전 1 14 46" src="https://github.com/user-attachments/assets/1094734c-8734-4ef4-ae88-4189e80c3d14">

Therefore, when the above code is processed by eslint scope-manager, no errors are reported because it meets the following conditions.

https://github.com/eslint/eslint/blob/main/lib/rules/no-useless-assignment.js#L311-L322

**Comment by yeonjuan** [comment]
I think I may have found the cause. ESLint assumes that all references are Identifiers.
So they update the last of the current segment to the last occurrence of the Identifier whenever it appears.
- https://github.com/eslint/eslint/blob/d46059410a0e02b98067aa31975c25fd8d0d1c2b/lib/rules/no-useless-assignment.js#L457-L466
```js
            Identifier(node) {
                for (const segment of scopeStack.currentSegments) {
                    const segmentInfo = scopeStack.segments[segment.id];


                    if (!segmentInfo.first) {
                        segmentInfo.first = node;
                    }
                    segmentInfo.last = node;
                }
            },
```

However, TSESLint Scope manager also creates a reference for JSXIdentifier. The JSXIdentifier reference generated by TSESLint is not used to update the segment of the rule.

That's why the return statement is not executed in this part.

- https://github.com/eslint/eslint/blob/main/lib/rules/no-useless-assignment.js#L383-L386
```js
                        // Uses in statements after the written identifier.
                        return;
                    }
```

 As new identifiers become available, segments are updated, which is why you see the following behavior
 
 ```tsx
 function App() {
  const AA = () => <></>; // This assigned value is not used in subsequent statements.
  const a = 1;
  return (
    <>
      {a}
      <AA>
      </AA>
    </>
  );
}
 ```

 ```tsx
 function App() {
  const AA = () => <></>; // no error
  const a = 1;
  return (
    <>
      <AA>
      </AA>
       {a} {/* Because a new Identifier has appeared. This expands the segment range.*/}
    </>
  );
}
 ```

**Comment by bradzacher** [comment]
This is complicated because you can't just naively ass `JSXIdentifier` to the selector because lower-case identifiers are not variable references. Additionally there are attribute names which also aren't references.

This would require a small amount of jsx specific logic - which is outside the scope of ESLint (they intentionally do not support JSX in rules)

**Comment by yeonjuan** [comment]
@bradzacher 
> This would require a small amount of jsx specific logic - which is outside the scope of ESLint (they intentionally do not support JSX in rules)

Yes, I agree. In my opinion, it might be a good idea to change ESLint to only check if the references are Indentifiers.
But do you still think this issue should be labeled “accepting PR”?

**Comment by bradzacher** [comment]
We could probably upstream a change to make the eslint rule only check Identifiers yeah. That seems like a good workaround. Someone should file an issue upstream.

**Comment by yeonjuan** [comment]
Sure https://github.com/eslint/eslint/issues/19160 👍

**Comment by yeonjuan** [comment]
It's fixed(https://github.com/eslint/eslint/pull/19200) and released (https://github.com/eslint/eslint/releases/tag/v9.17.0)
